packages/@aws-cdk/core/lib/custom-resource.ts

```diff
@@ -0,0 +1,196 @@
+import { CfnResource } from './cfn-resource';
+import { Construct } from './construct-compat';
+import { RemovalPolicy } from './removal-policy';
+import { Resource } from './resource';
+import { Token } from './token';
+
+/**
+ * Properties to provide a Lambda-backed custom resource
+ */
+export interface CustomResourceProps {
+  /**
+   * The ARN of the provider which implements this custom resource type.
+   *
+   * You can implement a provider by listening to raw AWS CloudFormation events
+   * and specify the ARN of an SNS topic (`topic.topicArn`) or the ARN of an AWS
+   * Lambda function (`lambda.functionArn`) or use the CDK's custom [resource
+   * provider framework] which makes it easier to implement robust providers.
+   *
+   * [resource provider framework]:
+   * https://docs.aws.amazon.com/cdk/api/latest/docs/custom-resources-readme.html
+   *
+   * Provider framework:
+   *
+   * ```ts
+   * // use the provider framework from aws-cdk/custom-resources:
+   * const provider = new custom_resources.Provider({
+   *   onEventHandler: myOnEventLambda,
+   *   isCompleteHandler: myIsCompleteLambda, // optional
+   * });
+   *
+   * new CustomResource(this, 'MyResource', {
+   *   serviceToken: provider.serviceToken
+   * });
+   * ```
+   *
+   * AWS Lambda function:
+   *
+   * ```ts
+   * // invoke an AWS Lambda function when a lifecycle event occurs:
+   * serviceToken: myFunction.functionArn
+   * ```
+   *
+   * SNS topic:
+   *
+   * ```ts
+   * // publish lifecycle events to an SNS topic:
+   * serviceToken: myTopic.topicArn
+   * ```
+   */
+  readonly serviceToken: string;
```

Comment:
Make this an `IServiceToken` or something with a single string attribute?

We can still have it typed, even if the dependency goes the other way. We can have `class Function implements IServiceToken`, for example.

Reply:
This will "pollute" the public APIs of `lambda.Function` and `sns.Topic` in an unwanted way, and I  don't think this is warranted for this use case. The documentation is clear and easy enough to follow and we expect people who implement custom resources to be able to easily follow it.

Reply:
If you think it's polluting Function then we could easily have migrated the `ResourceProvider` as well.

But fine, strings it is.

Reply:
> If you think it's polluting Function then we could easily have migrated the ResourceProvider as well.

How? `ResourceProvider.fromLambda` references a `IFunction` which does not exist in `core.